staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go

```diff
@@ -0,0 +1,550 @@
+
+import (
+	"fmt"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	xListTypeMap    = "map"
+	xListTypeAtomic = "atomic"
+	xListTypeSet    = "set"
+	xMapType        = "granular"
+
+	genericTypeInteger      = "integer"
+	genericTypeNumber       = "number"
+	genericTypeString       = "string"
+	genericTypeBool         = "boolean"
+	genericTypeObject       = "object"
+	genericTypeArray        = "array"
+	xImmutabilityImmutable  = "immutable"
+	xImmutabilityAddOnly    = "addOnly"
+	xImmutabilityRemoveOnly = "removeOnly"
+)
+
+// Immutable validates immutable values are respected or error otherwise
+func Immutable(old interface{}, new interface{}, s *structuralschema.Structural, pth *field.Path) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
+		}
+		for k, prop := range s.Properties {
+			// XImmutable and XImmutableKeys are mutually exclusive
+			// CRD validation should never let this be true
+			if prop.XImmutability != nil && prop.XKeyImmutability != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "XImmutability and XKeyImmutability are mutually exclusive"))
+				continue
+			}
+
+			_, foundOld := old[k]
+			_, foundNew := newObj[k]
+
+			// XImmutability
+			// validate immutable / addOnly /removeOnly against the current property regardless of the type.
+			// XImmutability/XKeyImmutability below this level is disallowed.
+			if prop.XImmutability != nil {
+				if *prop.XImmutability == xImmutabilityImmutable {
+					if (foundOld != foundNew) || !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityAddOnly {
+					if foundOld && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is addOnly"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityRemoveOnly {
+					if foundNew && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is removeOnly"))
+					}
+					continue
+				}
+				continue
+			}
+
+			// Mutable array, immutable items, list type undefined / atomic
+			// The items with their respective index in the array are immutable.
+			// Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+			if prop.Type == genericTypeArray &&
+				(derefPointerToString(prop.XListType) == "" || derefPointerToString(prop.XListType) == xListTypeAtomic) &&
```

Comment:
shouldn't we unify `atomic` and empty string in the structural types? I don't see a reason we keep this distinction in all this logic. They are the same thing. Simular to granular and empty for maps.